src/parser/validator.ts

```diff
@@ -0,0 +1,52 @@
+import { RootParser } from "../parser/RootParser";
+import { makeCstVisitor, VisitorResult, CommandNode } from "../parser/makeCstVisitor";
+
+interface validateCSTResult {
+    errors: ValidateCSTError[];
+    hasErrors: boolean;
+}
+
+interface ValidateCSTError {
+    error: Error,
+}
+
+export function validateVisitorResult(visitorResult: VisitorResult): validateCSTResult {
+    var validateResult = { errors: [], hasErrors: false};
+    var idSet = new Set<String>();
+    visitorResult.commands.forEach((command) => {
+        validateCST(command, validateResult, idSet);
+    });
+    return validateResult;
+}
+
+function validateCST(commandNode: CommandNode, errors: validateCSTResult, idSet: Set<String>) {
+    commandNode.children.forEach((child) => {
+        validateCST(child, errors, idSet);
+    });
+    codeBlockHasID(commandNode, errors);
+    idIsUnique(commandNode, errors, idSet);
+    return {
+        errors: errors,
+        hasErrors: errors.errors.length > 0
+    }
+}
+
+function codeBlockHasID(commandNode: CommandNode, errors: validateCSTResult) {
+    if (commandNode.commandName == "code-block") {
+        if (commandNode.id == undefined) {
+            errors.errors.push({error: {name: "code-block ID missing", message: "line " + commandNode.range.start.line + ": missing ID on a code block"}});
```

Comment:
Prefer template literals:
```suggestion
            errors.errors.push({error: {name: "code-block ID missing", message: `line ${commandNode.range.start.line}: missing ID on a code block`}});
```

Reply:
Because I'm using VisitorError now, no need to embed the location in the message itself.

Reply:
The point was to use template literals rather than concatenating strings ("this is x: " + x + " that was x"). String concatenation like most things JS has weird behavior and it is also less legible.

Reply:
Understood. I just meant that the line now looks like this:
```
result.errors.push({location: commandNode.range.start, message: "duplicate ID on a code block"});
```
so we don't need to use template literals *or* concatenation.